types/stampit/index.d.ts

```diff
@@ -1,518 +1,560 @@
-// Type definitions for stampit 3.0
+// Type definitions for stampit 4.3
 // Project: https://github.com/stampit-org/stampit, https://stampit.js.org
 // Definitions by: Vasyl Boroviak <https://github.com/koresar>
 //                 Harris Lummis <https://github.com/lummish>
+//                 PopGoesTheWza <https://github.com/popgoesthewza>
 // Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+// TypeScript Version: 3.5
+
+// Utility types
+
+/**
+ * @internal Base type for all `methods`-like metadata.
+ * @template This The type to use for `this` within methods.
+ */
+interface MethodMap<This> {
+    [s: string]: (this: This, ...args: any[]) => any;
+}
+
+/** @internal A plain old JavaScript object created by a `Stamp`. */
+type Pojo = object; // { [s: string]: any; }
+
+/** @internal Base type for all `properties`-like metadata. */
+// TODO: discriminate Array
+type PropertyMap = object; // { [s: string]: any; }
+
+/** @internal Signature common to every `Stamp`s. */
+interface StampSignature {
+    (options?: PropertyMap, ...args: unknown[]): any;
+    compose: ComposeMethod & stampit.Descriptor<any, any>;
+}
+
+/**
+ * @internal Extracts the `Stamp` type.
+ * @template Original The type to extract the `Stamp` type from.
+ */
+type StampType<Original> = Original extends /* disallowed types */ [] | bigint
```

Comment:
This whole `type` is a black magic to me. :) I'll pretend it's all correct. Although, the `[]`, `bigint`, `true` look weird.

Reply:
It is indeed rather black magic for it will test all cases I could think of and try to detect if and how a stamp type can be inferred: it works out of Descriptor and regular Stamp types. It worked with “extended” Stamp types as far as ì could test to (cf. the test .ts file)
The initial type test is to properly return the never type for types which would otherwise produce (wrongfully) a valid return types.
This use of TypeScript types is not typical. It attempts to cover as many uses of stamps as possible without requiring too formal type definition upfront.